test/tools_tests/tree_metrics_test.py

```diff
@@ -0,0 +1,414 @@
+"""
+Tests for cassiopeia/tools/tree_metrics.py
+"""
+import unittest
+from typing import Dict, Optional
+
+import itertools
+import networkx as nx
+import numpy as np
+import pandas as pd
+
+import cassiopeia as cas
+import cassiopeia.tools.tree_metrics as tm
+from cassiopeia.data.CassiopeiaTree import (
+    CassiopeiaTreeError,
+    CassiopeiaTreeWarning,
+)
+
+class TestCassiopeiaTree(unittest.TestCase):
+    def setUp(self):
+        small_net = nx.DiGraph()
+        small_net.add_edges_from(
+            [
+                ("node5", "node0"),
+                ("node5", "node1"),
+                ("node6", "node2"),
+                ("node6", "node3"),
+                ("node6", "node4"),
+                ("node7", "node5"),
+                ("node7", "node6"),
+            ]
+        )
+        self.small_net = small_net
+
+
+    def test_parsimony_bad_cases(self):
+        small_tree = cas.data.CassiopeiaTree(tree=self.small_net)
+        with self.assertRaises(CassiopeiaTreeError):
+            tm.calculate_parsimony(small_tree, infer_ancestral_characters=False)
+
+        small_cm = pd.DataFrame.from_dict(
+            {
+                "node0": [1, -1, -1],
+                "node1": [2, 1, -1],
+                "node2": [2, -1, -1],
+                "node3": [1, 2, 2],
+                "node4": [1, 1, 2],
+            },
+            orient="index",
+        )
+        small_tree = cas.data.CassiopeiaTree(
+            tree=self.small_net, character_matrix=small_cm
+        )
+        with self.assertRaises(CassiopeiaTreeError):
+            tm.calculate_parsimony(small_tree, infer_ancestral_characters=False)
+
+        small_tree.set_character_states("node7", [0, 0, 0])
+        small_tree.set_character_states("node5", [0, 0, 0])
+        with self.assertRaises(CassiopeiaTreeError):
+            tm.calculate_parsimony(small_tree, infer_ancestral_characters=False)
+
+    def test_parsimony_reconstruct_internal_states(self):
+        small_cm = pd.DataFrame.from_dict(
```

Comment:
it appears that this character matrix is used several times throughout the tests, it would be a lot cleaner if you made this a part of the test class in the `setUp()` function.